commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt

```diff
@@ -0,0 +1,190 @@
+package com.github.quillraven.commons.audio
+
+import com.badlogic.gdx.audio.Music
+import com.badlogic.gdx.audio.Sound
+import com.badlogic.gdx.utils.ObjectMap
+import com.badlogic.gdx.utils.Pool
+import kotlinx.coroutines.launch
+import ktx.assets.async.AssetStorage
+import ktx.async.KtxAsync
+import ktx.collections.getOrPut
+import ktx.collections.iterate
+import ktx.collections.set
+import ktx.log.debug
+import ktx.log.error
+import ktx.log.info
+import ktx.log.logger
+import kotlin.math.max
+
+/**
+ * A request for a [Sound] that is created within the [QueueAudioService.playSound] method.
+ * It contains the [filePath] and the [volume] of the request.
+ */
+private class SoundRequest : Pool.Poolable {
+  var filePath: String = ""
+  var volume = 1f
+
+  override fun reset() {
+    filePath = ""
+    volume = 1f
+  }
+}
+
+/**
+ * Implementation of [AudioService] to play one [Music] track at a time and to queue [Sound] playback
+ * to avoid multiple [Sound] instances of the same sound within a single frame.
+ *
+ * Requires an [assetStorage] to load and unload the [Sound] and [Music] resources.
+ *
+ * Use [maxSimultaneousSounds] to define the maximum number of sounds that can play concurrently within one frame.
+ *
+ * Use [maxCachedSounds] to define the size of the sound cache. [Sound] instances are cached internally for
+ * faster access of future calls. If the [maxCachedSounds] is reached then the cache gets cleared.
+ */
+class QueueAudioService(
+  private val assetStorage: AssetStorage,
+  private val maxSimultaneousSounds: Int = 16,
+  private val maxCachedSounds: Int = 50
+) : AudioService {
+
+  private val soundRequests = ObjectMap<String, SoundRequest>(maxSimultaneousSounds)
```

Comment:
Is it not possible to play the same sound at the same time? I'd imagine sounds overlap would be pretty common depending on the game type. Jumping, shooting, or even more complex particles could play overlapping sounds at more or less the same time.

Reply:
From what I know if you play 2 sounds in the same frame the volume will be doubled, 3 times will be tripple the volume, etc... In my opinion this does not make sense and should be prevented to not explode the ears of the gamers ;)

Reply:
@Quillraven But you have to consider overlapping sounds, e.g. when you play a sound in the middle of the previous one. You might want to prevent playing the same sound in the same exact frame, but if there's any kind of delay, I think that's a valid use case. For example, this might occur if you're firing a weapon multiple times, double-jumping or casting a spell with multiple projectiles. This will be especially noticeable in case of the longer sounds.

Reply:
That should still be possible because the `update` method should be called in each `render` loop and then the requests get processed and cleared and new ones can be added. Therefore, it should only avoid sound overlaps within the exact same frame. If it is a frame later, it will still be played.

Reply:
Ah, I see.